src/googleclouddebugger/firebase_client.py

```diff
@@ -0,0 +1,571 @@
+    """Sets up authentication with Google APIs.
+
+    This will use the credentials from service_account_json_file if provided,
+    falling back to application default credentials.
+    See https://cloud.google.com/docs/authentication/production.
+
+    Args:
+      project_id: GCP project ID (e.g. myproject). If not provided, will attempt
+          to retrieve it from the credentials.
+      service_account_json_file: JSON file to use for credentials. If not
+          provided, will default to application default credentials.
+    Raises:
+      NoProjectIdError: If the project id cannot be determined.
+    """
+    if service_account_json_file:
+      self._credentials = credentials.Certificate(service_account_json_file)
+      if not project_id:
+        with open(service_account_json_file, encoding='utf-8') as f:
+          project_id = json.load(f).get('project_id')
+    else:
+      if not project_id:
+        try:
+          r = requests.get(
+              f'{_METADATA_SERVER_URL}/project/project-id',
+              headers={'Metadata-Flavor': 'Google'})
+          # TODO: Check whether more needs to be done here.
+          project_id = r.text
+        except requests.exceptions.RequestException:
+          native.LogInfo('Metadata server not available')
+
+    if not project_id:
+      raise NoProjectIdError(
+          'Unable to determine the project id from the API credentials. '
+          'Please specify the project id using the --project_id flag.')
+
+    self._project_id = project_id
+    # TODO: Add database url override support.
+    self._database_url = f'https://{self._project_id}-cdbg.firebaseio.com'
+
+  def Start(self):
+    """Starts the worker thread."""
+    self._shutdown = False
+
+    # TODO: Sort out what needs to be done here.
+    self._main_thread = threading.Thread(target=self._MainThreadProc)
+    self._main_thread.name = 'Cloud Debugger main worker thread'
+    self._main_thread.daemon = True
+    self._main_thread.start()
+
+  def Stop(self):
+    """Signals the worker threads to shut down and waits until it exits."""
+    self._shutdown = True
+    self._new_updates.set()  # Wake up the transmission thread.
+
+    if self._main_thread is not None:
+      self._main_thread.join()
+      self._main_thread = None
+
+    if self._transmission_thread is not None:
+      self._transmission_thread.join()
+      self._transmission_thread = None
+
+    if self._breakpoint_subscription is not None:
+      self._breakpoint_subscription.close()
+      self._breakpoint_subscription = None
+
+  def EnqueueBreakpointUpdate(self, breakpoint_data):
+    """Asynchronously updates the specified breakpoint on the backend.
+
+    This function returns immediately. The worker thread is actually doing
+    all the work. The worker thread is responsible to retry the transmission
+    in case of transient errors.
+
+    The assumption is that the breakpoint is moving from Active to Final state.
```

Comment:
Python agent does not support updating the breakpoint line number if no code was found on the requested line?